### PR TITLE
Add schema validation

### PR DESCRIPTION
**Pull request by bugagashenkj**

### CODE REVIEW

File: test/non-serializable.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    test.strictSame(
+      error.toString(),
+      new MetaschemaError([
+        new SchemaValidationError('nonSerializable', 'test.func'),
+        new SchemaValidationError('nonSerializable', 'Test.Func.check'),
+        new SchemaValidationError('nonSerializable', 'Action.Args.Func.check'),
+      ]).toString()
+    );
+    test.end();
+    return;
```
Comment (belochub) [inline test/non-serializable.js]: Why did you add a `return` here?

File: test/non-serializable.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+const metaschema = require('metaschema');
+const metatests = require('metatests');
+
+const { SchemaValidationError, MetaschemaError } = metaschema.errors;
+const { options, config } = require('../lib/metaschema-config/config');
+
+metatests.test('Non-serializable functions', async test => {
+  try {
+    await metaschema.fs.load('test/fixtures/non-serializable', options, config);
+    test.fail();
```
Comment (belochub) [inline test/non-serializable.js]: IIRC, `test.fail()` requires a message. Also, you can replace it with `test.rejects()` and remove `try/catch` with the next `metatests` release.

File: test/non-serializable.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  } catch (error) {
+    test.strictSame(
+      error.toString(),
+      new MetaschemaError([
+        new SchemaValidationError('nonSerializable', 'test.func'),
+        new SchemaValidationError('nonSerializable', 'Test.Func.check'),
+        new SchemaValidationError('nonSerializable', 'Action.Args.Func.check'),
+      ]).toString()
+    );
+    test.end();
```
Comment (belochub) [inline test/non-serializable.js]: There is no need to call `test.end()` since you are using an async function.

File: lib/metaschema-config/config.js [hunk 4]
```diff
@@ -84,6 +87,29 @@ const preprocessor = schemas => {
   return arr;
 };
 
+const validate = (schema, path) => {
+  const errors = [];
+
+  for (const [prop, value] of Object.entries(schema)) {
+    const valueType = typeof value;
+    if (valueType === 'object') {
+      errors.push(...validate(value, `${path}.${prop}`, errors));
```
Comment (belochub) [inline lib/metaschema-config/config.js]: The function only accepts 2 arguments, why do you pass `errors` as a third argument?

File: lib/metaschema-config/config.js [hunk 5]
```diff
@@ -84,6 +87,29 @@ const preprocessor = schemas => {
...
+    } else if (valueType === 'function') {
```
Comment (nechaido) [inline lib/metaschema-config/config.js]: This will not work for `class` and `Class` fields in schemas.

**Comment by bugagashenkj** [comment]
@nechaido

File: test/fixtures/non-serializable/test.domains [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+  func: Date.parse
```
Comment (nechaido) [inline test/fixtures/non-serializable/test.domains]:
```suggestion
  func: { type: 'object', class: 'Date', parse: Date.parse }
```

File: lib/metaschema-config/config.js [hunk 7]
```diff
@@ -84,6 +87,46 @@ const preprocessor = schemas => {
...
+      const stringifyValue = value.toString();
+      let err;
+      try {
+        new Function(`(${stringifyValue})`);
+      } catch (e) {
+        err = e;
+      }
+
+      if (err && !value.prototype) {
```
Comment (lundibundi) [inline lib/metaschema-config/config.js]: Could you please add comments to note exactly why there are two `try-catch` and why this `value.prototype` is relevant?

File: test/non-serializable.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+metatests.test('Non-serializable functions', async test => {
+  const error = await test.rejects(
+    metaschema.fs.load('test/fixtures/non-serializable', options, config)
+  );
+  test.strictSame(
+    error,
+    new MetaschemaError([
+      new SchemaValidationError('nonSerializable', 'test.func.parse'),
+      new SchemaValidationError('nonSerializable', 'Test.Func1.check'),
+      new SchemaValidationError('nonSerializable', 'Action.Args.Func.check'),
```
Comment (lundibundi) [inline test/non-serializable.js]: Isn't this case missing category name in path?

**Comment by belochub** [comment]
Landed in https://github.com/metarhia/globalstorage/commit/667d68842f53ab3fbf27b0e387958382a675d57e.
